### PR TITLE
fix: npm publish verification

### PR DESCRIPTION
**Pull request by laurentsimon**
- adding support for IEEE P1363 formatted signatures
- fix the npm publish attestation bug. The verification always return success, because it was not using PAE signature

### CODE REVIEW

**Comment by laurentsimon** [comment]
@trishankatdatadog PTAL

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:45]: nit: be consistent with `EnvelopeFromBytes`
```suggestion
func StatementFromBytes(payload []byte) (*intoto.Statement, error) {
```

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:42]: nit: on line 39 change to

```golang
if len(payload) == 0 {
```

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:52]: nit: error message
```suggestion
		return nil, fmt.Errorf("%w: invalid statement type: %q", serrors.ErrorInvalidDssePayload, statement.Type)
```

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:48]: nit: wrap both errors
```suggestion
		return nil, fmt.Errorf("%w: %w", serrors.ErrorInvalidDssePayload, err)
```

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:28]: nit: formatting
```suggestion
		return nil, fmt.Errorf("%w: expected payload type %q, got %q",
```

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:24]: nit: wrap both errors
```suggestion
		return nil, fmt.Errorf("%w: %w", serrors.ErrorInvalidDssePayload, err)
```

File: verifiers/internal/gha/verifier.go
Comment (ianlewis) [inline verifiers/internal/gha/verifier.go:134]: nit: "empty" by itself is a bit vague
```suggestion
		return nil, fmt.Errorf("%w: no expected builder ID", serrors.ErrorInvalidBuilderID)
```

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:172]: nit: not sure we should include the entire public key contents in the error message.
```suggestion
			return nil, fmt.Errorf("%w: unable to decode PEM format", serrors.ErrorInvalidPEM)
```

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:170]: Should we check if `rest` has some data in it? What if there are multiple PEM blocks? Should we try to check all of them?

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:179]: nit: Maybe we can have an `serrors.InvalidPublicKey` error?

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:195]: nit: error message
```suggestion
		return nil, fmt.Errorf("creating verifier: %w", err)
```

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:202]: nit: You could pass signature encoding to `DsseVerifierNew` instead. That way you could just return the `dsselib.EnvelopeVerifier` instead of having a separate `DsseVerifier` type and `Verify` method.

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:106]: nit: This is maybe more [Go-like](https://go.dev/doc/effective_go#mixed-caps) but you could also add a [golangci-lint false positive comment](https://golangci-lint.run/usage/false-positives/).
```suggestion
	SignatureEncodingIEEEP1363
```

File: verifiers/internal/gha/npm.go
Comment (ianlewis) [inline verifiers/internal/gha/npm.go:315]: nit: wrap both errors
```suggestion
		return nil, fmt.Errorf("%w: %w", serrors.ErrorInvalidDssePayload, err)
```

File: verifiers/internal/gha/npm.go
Comment (ianlewis) [inline verifiers/internal/gha/npm.go:321]: nit: wrap both errors
```suggestion
		return nil, fmt.Errorf("%w: %w", serrors.ErrorInvalidDssePayload, err)
```

File: verifiers/internal/gha/npm.go
Comment (ianlewis) [inline verifiers/internal/gha/npm.go:325]: nit: msg formatting
```suggestion
		return nil, fmt.Errorf("%w: no subjects", serrors.ErrorInvalidDssePayload)
```

File: verifiers/internal/gha/npm.go
Comment (ianlewis) [inline verifiers/internal/gha/npm.go:396]: nit: formatting
```suggestion
	return fmt.Errorf("expected hash %q not found: %w", expectedHash, serrors.ErrorMismatchHash)
```

File: verifiers/internal/gha/npm.go
Comment (ianlewis) [inline verifiers/internal/gha/npm.go:380]: nit: I think this is a more understandable error message.
```suggestion
		return fmt.Errorf("%w: expected minimum sha256, got %s", serrors.ErrorInvalidHash, expectedAlgo)
```

**Comment by ianlewis** [comment]
I like that you used the `dsselib.EnvelopeVerifier` interface.

Could you maybe add a bit more in the PR description as to what this PR is doing? esp. that this is adding support for IEEE P1363 formatted signatures.

**Comment by trishankatdatadog** [comment]
Can confirm the patch works:

```shell
➜  slsa-verifier git:(fix/npm-publish-sig) SLSA_VERIFIER_EXPERIMENTAL=1 go run ./cli/slsa-verifier verify-npm-package ~/GitHub.com/trishankatdatadog/hekate/supreme-goggles.tgz \
  --attestations-path ~/GitHub.com/trishankatdatadog/hekate/attestations.json \
  --builder-id "https://github.com/actions/runner/github-hosted" \
  --package-name "@trishankatdatadog/supreme-goggles" \
  --package-version 1.0.5 \
  --source-uri github.com/trishankatdatadog/supreme-goggles
Verified build using builder https://github.com/actions/runner/github-hosted at commit 38ebf99444e033b2f1550c9aaaeacd62d02a12ba
Verifying npm package /Users/trishank.kuppusamy/GitHub.com/trishankatdatadog/hekate/supreme-goggles.tgz: FAILED: invalid signature: accepted signatures do not match threshold, Found: 0, Expected 1

FAILED: SLSA verification failed: invalid signature: accepted signatures do not match threshold, Found: 0, Expected 1
exit status 1
```

Will try to review code ASAP. Thanks for fixing this!

File: verifiers/utils/dsse.go
Comment (laurentsimon) [reply verifiers/utils/dsse.go:179]: great idea. done.

File: verifiers/utils/dsse.go
Comment (laurentsimon) [reply verifiers/utils/dsse.go:202]: thanks foe the suggestion. Initially I wanted to decouple signature verification (and its encoding) from the verification public key (you can technically verify signatures with different encoding with the same public key). But in our case it's not really needed. Updated as you suggested.

File: cli/slsa-verifier/main_regression_test.go
Comment (trishankatdatadog) [inline cli/slsa-verifier/main_regression_test.go:1590]:
```suggestion
			pkgName:   PointerTo("@trishankatdatadog/supreme-goggles"),
```

File: cli/slsa-verifier/main_regression_test.go
Comment (laurentsimon) [reply cli/slsa-verifier/main_regression_test.go:1499]: I've added tests (invalidsig*) that tests when verification fails. I updated a field in the attestation. Is that OK?

File: cli/slsa-verifier/main_regression_test.go
Comment (laurentsimon) [reply cli/slsa-verifier/main_regression_test.go:1590]: good catch, thank you!

File: verifiers/internal/gha/npm.go
Comment (laurentsimon) [reply verifiers/internal/gha/npm.go:49]: yes we can, there's a tracking issue for it. But we have not implemented yet.

Comment (laurentsimon) [reply verifiers/internal/gha/npm.go:49]: here https://github.com/slsa-framework/slsa-verifier/issues/547

File: verifiers/internal/gha/npm.go
Comment (laurentsimon) [reply verifiers/internal/gha/npm.go:140]: they dont. provenance verification uses cosign API. npm verification overall is in https://github.com/slsa-framework/slsa-verifier/blob/main/verifiers/internal/gha/verifier.go#L308-L362.
verifyBuilderID is https://github.com/slsa-framework/slsa-verifier/blob/main/verifiers/internal/gha/npm.go#L297, which ends up calling cosign API

File: cli/slsa-verifier/main_regression_test.go
Comment (trishankatdatadog) [reply cli/slsa-verifier/main_regression_test.go:1499]: Oh, I see, the invalidsig tests would (predictably) fail because the subject digests could have been tampered with. WFM.

File: verifiers/internal/gha/npm.go
Comment (trishankatdatadog) [reply verifiers/internal/gha/npm.go:140]: OK, let's worry about code reuse in the future. Thx!

**Review by trishankatdatadog — APPROVED** [review]
LGTM, could use another pair of eyes to see whether I missed anything, thx!

File: cli/slsa-verifier/main_regression_test.go
Comment (laurentsimon) [reply cli/slsa-verifier/main_regression_test.go:1499]: correct. I have not altered the digest, just some other fields, not the digest. It's the same check more or less. Resolving. Re-open if disagree, thanks!

**Comment by laurentsimon** [comment]
> LGTM, could use another pair of eyes to see whether I missed anything, thx!

@ianlewis provided a good review already. Waiting for his LGTM too. Thanks again!

**Comment by laurentsimon** [comment]
> Just to double-check: we do verify that the _observed_ artefact digest does match the _expected_ subject digest in _both_ the provenance and publish attestations, correct?

yes

**Comment by trishankatdatadog** [comment]
> yes

Sorry to press you, but is there a test to this effect? I may have missed it...

File: cli/slsa-verifier/main_regression_test.go
Comment (ianlewis) [inline cli/slsa-verifier/main_regression_test.go:1682]: We have tests for package name, version, and signature mismatches but @trishankatdatadog mentioned that we are missing a test for a publish attestation digest mismatch.

File: verifiers/utils/dsse.go
Comment (ianlewis) [inline verifiers/utils/dsse.go:166]: nit
```suggestion
		if len(rest) != 0 {
```

**Review by ianlewis — APPROVED** [review]
LGTM

**Comment by laurentsimon** [comment]
> > yes
> 
> Sorry to press you, but is there a test to this effect? I may have missed it...

the tests are not part of this PR. They are present in the rest of the code.

File: cli/slsa-verifier/main_regression_test.go
Comment (trishankatdatadog) [reply cli/slsa-verifier/main_regression_test.go:1682]: Should we create an issue to track this?

Comment (laurentsimon) [reply cli/slsa-verifier/main_regression_test.go:1682]: is https://github.com/slsa-framework/slsa-verifier/issues/707 ok?

**Pull request merged**
